config/cmake/modules/FindHYPRE.cmake

```
@@ -28,6 +31,55 @@ if (HYPRE_FOUND)
   return()
 endif()
 
+if (HYPRE_FETCH OR FETCH_TPLS)
+  set(HYPRE_FETCH_VERSION 2.32.0)
+  add_library(HYPRE STATIC IMPORTED)
+  # set options and associated dependencies
+  set(CMAKE_OPTIONS)
+  list(APPEND CMAKE_OPTIONS -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE})
+  if (MFEM_USE_CUDA)
+    list(APPEND CMAKE_OPTIONS -DHYPRE_WITH_CUDA:BOOL=ON)
+    find_package(CUDAToolkit REQUIRED)
+    target_link_libraries(HYPRE INTERFACE CUDA::cusparse CUDA::curand CUDA::cublas)
+  elseif (MFEM_USE_HIP)
+    list(APPEND CMAKE_OPTIONS -DHYPRE_WITH_HIP:BOOL=ON)
+    find_package(rocsparse REQUIRED)
+    find_package(rocrand REQUIRED)
+    target_link_libraries(HYPRE INTERFACE rocsparse rocrand)
+  endif()
+  if (MFEM_USE_SINGLE)
+    list(APPEND CMAKE_OPTIONS -DHYPRE_ENABLE_SINGLE:BOOL=ON)
+  endif()
+  # define external project and create future include directory so it is present
+  # to pass CMake checks at end of MFEM configuration step
+  message(STATUS "Will fetch HYPRE ${HYPRE_FETCH_VERSION} to be built with ${CMAKE_OPTIONS}")
+  set(PREFIX ${CMAKE_BINARY_DIR}/fetch/hypre)
+  include(ExternalProject)
+  ExternalProject_Add(hypre
+    GIT_REPOSITORY https://github.com/hypre-space/hypre.git
+    GIT_TAG v${HYPRE_FETCH_VERSION}
+    GIT_SHALLOW TRUE
+    UPDATE_DISCONNECTED TRUE
+    SOURCE_SUBDIR src
+    PREFIX ${PREFIX}
+    CMAKE_CACHE_ARGS -DCMAKE_INSTALL_PREFIX:PATH=${PREFIX} -DCMAKE_INSTALL_LIBDIR:PATH=lib ${CMAKE_OPTIONS})
+  file(MAKE_DIRECTORY ${PREFIX}/include)
+  # set imported library target properties
+  add_dependencies(HYPRE hypre hypre-install)
```

`hypre-install` doesn't seem to have any effect (I tried renaming, removing, no change either way)

```suggestion
  add_dependencies(HYPRE hypre)
```

---

`hypre-install` should be a reference to the ExternalProject "install" step of hypre.  I found this was necessary in parallel builds to prevent, e.g., an mfem file that includes a hypre header to be compiled before that hypre header exists.  I have done the same for metis, i.e., the `METIS` target depends on the completion of the "install" step for metis.

---

Upon revisiting some larger parallel builds, it looks like the race condition I needed that extra dependency for was addressed when `UPDATE_DISCONNECTED TRUE` was added to the hypre ExternalProject configuration.. I'll remove these depedencies (which I suppose should have been added using `ExternalProject_Add_StepDepedencies` anyways)

---

> hypre-install should be a reference to the ExternalProject "install" step of hypre.

Do you have a source for this?

> I found this was necessary in parallel builds to prevent, e.g., an mfem file that includes a hypre header to be compiled before that hypre header exists.

Strange, for me the `hypre` dependency does exactly that. I've written a check like below, and it doesn't print no matter the status of `UPDATE_DISCONNECTED`:
```
if (TARGET hypre-install)
  message(STATUS "HELLO")
endif()
```

> I have done the same for metis

Well noted, I'll request the same change for METIS as well just for our reference.

---

I'm removing the dependencies on the install step as you requested... I don't see any further discussion necessary?

---

See the CMake documentation on ExternalProject for the step dependencies.  I'm not sure they were fully implemented correctly, which will be addressed when I remove them as the problem I put any implementation in to address no longer exists.

---

I apologise, I just like to get to the bottom of things. Thank you for accepting the suggestions.